src/component/handlers/composite-android/DraftEditorCompositionHandlerAndroid.js

```diff
@@ -0,0 +1,198 @@
+/**
+ * Copyright (c) 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule DraftEditorCompositionHandlerAndroid
+ * @typechecks
+ * @flow
+ */
+
+'use strict';
+
+import type DraftEditor from 'DraftEditor.react';
+
+const DraftModifier = require('DraftModifier');
+const EditorState = require('EditorState');
+const ReactDOM = require('ReactDOM');
+
+const getDraftEditorSelection = require('getDraftEditorSelection');
+
+let compositionRange = undefined;
+let compositionText = undefined;
+let hasInsertedCompositionText = false;
+let hasMutation = false;
+
+const resetCompositionData = () => {
+  createMutationObserverIfUndefined();
+  compositionRange = undefined;
+  compositionText = undefined;
+  hasInsertedCompositionText = false;
+  hasMutation = false;
+};
+
+const handleMutations = records => {
+  hasMutation = !!records.length;
+  mutationObserver.disconnect();
+};
+
+let mutationObserver = undefined;
+
+// Need to create mutation observer on runtime, not compile time
+const createMutationObserverIfUndefined = () => {
+  if (!mutationObserver) {
+    mutationObserver = new window.MutationObserver(handleMutations);
+  }
+};
+
+/**
+ * Replace the current selection with the specified text string, with the
+ * inline style and entity key applied to the newly inserted text.
+ */
+function replaceText(
+    editorState: EditorState,
+    text: string,
+    compositionRange: SelectionState,
+    // inlineStyle: DraftInlineStyle,
+    // entityKey: ?string,
+  ): EditorState {
+  var contentState = DraftModifier.replaceText(
+      editorState.getCurrentContent(),
+      compositionRange,
+      text,
+    //   inlineStyle,
+    //   entityKey,
+    );
+  return EditorState.push(editorState, contentState, 'insert-characters');
+}
+
+const getCompositionRange = (editor, text) => {
+  // Ocassionally a newline will get composed.  In this case, we want to strip it since
+  // we won't be able to match it in Draft, and it will get rewritten anyways.
+  if (text.endsWith('\n')) {
+    text = text.slice(0, text.length - 1);
+  }
+
+  if (!text) {
+    // get Selection (Assuming editorState is correct…)
+    // Since we know editorState is often out of sync right now, derive from the DOM:
+    const editorNode = ReactDOM.findDOMNode(editor.refs.editorContainer);
```

Comment:
< 1 :star: It seems like the combination of this line & the next show up together a lot and could be pulled out into a helper for readability.

Reply:
In the very short term we can just pull it up out of the conditional since we do it in either scenario.
That said, we can also do a cleanup pass in a later PR.

Reply:
I'll save that for another refactor if that's cool